rust/noosphere-sphere/src/walker.rs

```diff
@@ -28,31 +30,153 @@ where
     K: KeyMaterial + Clone + 'static,
     S: Storage + 'static,
 {
     fn from(has_sphere_context: H) -> Self {
-        SphereContentWalker {
+        SphereWalker {
             has_sphere_context,
             key: Default::default(),
             storage: Default::default(),
         }
     }
 }
 
-impl<H, K, S> SphereContentWalker<H, K, S>
+impl<H, K, S> SphereWalker<H, K, S>
+where
+    H: SpherePetnameRead<K, S> + HasSphereContext<K, S>,
+    K: KeyMaterial + Clone + 'static,
+    S: Storage + 'static,
+{
+    // Get a stream that yields every petname in the namespace along with its
+    /// corresponding [AddressIpld]. This is useful for iterating over sphere
+    /// petnames incrementally without having to load the entire index into
+    /// memory at once.
+    pub fn petname_stream<'a>(&'a self) -> impl Stream<Item = Result<(String, AddressIpld)>> + 'a {
+        try_stream! {
+            let sphere = self.has_sphere_context.to_sphere().await?;
+            let petnames = sphere.get_names().await?;
+            let stream = petnames.stream().await?;
+
+            for await entry in stream {
+                let (petname, address) = entry?;
+                yield (petname.clone(), address.clone());
+            }
+        }
+    }
+
+    /// Get a stream that yields the set of petnames that changed at each
+    /// revision of the backing sphere, up to but excluding an optional `since`
+    /// CID parameter. To stream the entire history, pass `None` as the
+    /// parameter.
+    pub fn petname_change_stream<'a>(
+        &'a self,
+        since: Option<&'a Cid>,
+    ) -> impl Stream<Item = Result<(Cid, BTreeSet<String>)>> + 'a {
+        try_stream! {
+            let sphere = self.has_sphere_context.to_sphere().await?;
+            let since = since.cloned();
+            let stream = sphere.into_name_changelog_stream(since.as_ref());
+
+            for await change in stream {
+                let (cid, changelog) = change?;
+                let mut changed_petnames = BTreeSet::new();
+
+                for operation in changelog.changes {
+                    let petname = match operation {
+                        MapOperation::Add { key, .. } => key,
+                        MapOperation::Remove { key } => key,
+                    };
+                    changed_petnames.insert(petname);
+                }
+
+                yield (cid, changed_petnames);
+            }
+        }
+    }
+
+    /// Get a [BTreeSet] whose members are all the petnames that have addresses
+    /// as of this version of the sphere. Note that the full space of names may
+    /// be very large; for a more space-efficient approach, use
+    /// [SphereWalker::petname_stream] to incrementally access all petnames in
+    /// the sphere.
+    ///
+    /// This method is forgiving of missing or corrupted data, and will yield an
+    /// incomplete set of names in the case that some or all names are not able
+    /// to be accessed.
```

Comment:
Question: is there a specific cause of missing data / corruption in mind here that would result in failing to read a petname?

Reply:
A very common cause may be: you haven't fully replicated the sphere you are reading from yet. Ideally, this is a protocol detail that should not leak to the caller, but in practice I'm certain we will hit this from time to time.

IPLD is nice because it enables us to incrementally load a sphere when we want to read a narrow slice of it. But, one drawback of this is: just because you have part of a sphere doesn't mean you have the whole sphere. If you load part of the address book (say, enough to read the key space, but not all the values), and then go offline, you may not be able to read parts of the address book that you are iterating over.

There is probably a UX to be explored where we indicate to the reader that the sphere is in an indeterminate state (due to only being partially available), to better communicate what is happening in cases like this.